[iOS] Allow WKWebView clients to customize the file upload flow for file inputs

lilyspiniolas opened the pull request:
#### 0b662f0233021039f359f91308fe4762e78d1f97
<pre>
[iOS] Allow WKWebView clients to customize the file upload flow for file inputs
<a href="https://bugs.webkit.org/show_bug.cgi?id=280537">https://bugs.webkit.org/show_bug.cgi?id=280537</a>
<a href="https://rdar.apple.com/130219174">rdar://130219174</a>

Reviewed by Wenson Hsieh.

These changes allow clients to implement &apos;webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:&apos;
on iOS if they desire non-default behavior for the file upload flow in
WKWebView. These changes also make WKFileUploadPanel&apos;s
&apos;_copyToNewTemporaryDirectory&apos; a class method so that the fix for
<a href="https://rdar.apple.com/78448610">rdar://78448610</a> can be reused here without code duplication.

* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParameters.h:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParameters.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParametersInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParametersPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::runOpenPanel):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(TestWebKitAPI::TEST(WebKit, RunOpenPanelNonLatin1)):
(TestWebKitAPI::TEST(WebKit, FileInputTypeCancelEvent)):

Canonical link: <a href="https://commits.webkit.org/284471@main">https://commits.webkit.org/284471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/455859360bb183b4ac7049a6609f6e29176126be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48894 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73578 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71611 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55260 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13723 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72560 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60002 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17432 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63207 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17778 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75289 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13476 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62928 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62837 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15431 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4482 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44698 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45772 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45513 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -844,6 +844,50 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
+        Vector<String> filenames;
+#if PLATFORM(IOS)
+        RetainPtr<NSFileCoordinator> uploadFileCoordinator = [[NSFileCoordinator alloc] init];
+        RetainPtr<NSFileManager> uploadFileManager = [[NSFileManager alloc] init];
```
I think these need adoptNS, right?

pxlcoder replied:
Yes, this is a memory leak as-is.

rr-codes commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -844,6 +844,50 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
+
+    auto frame = API::FrameInfo::create(WTFMove(frameInfo), &page);
+
+    auto checker = CompletionHandlerCallChecker::create(delegate.get(), @selector(webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:));
```
If these return Ref or RefPtrs, they should use that instead of auto

lilyspiniolas replied:
Those were using auto before my changes; is it worth updating anyways? @rr-codes

pxlcoder replied:
Yeah, `auto` -> explicit smart pointer is a nice change to make when moving code.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
```diff
@@ -52,6 +52,7 @@ enum class PickerDismissalReason : uint8_t;
 
 - (NSArray<NSString *> *)currentAvailableActionTitles;
 - (NSArray<NSString *> *)acceptedTypeIdentifiers;
++ (std::pair<RetainPtr<NSURL>, RetainPtr<NSURL>>)_copyToNewTemporaryDirectory:(NSURL *)originalURL fileCoordinator:(NSFileCoordinator *)fileCoordinator fileManager:(NSFileManager *)fileManager;
```
Nit: Add a newline before this declaration.

whsieh commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -844,6 +844,50 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
+    if (!delegate)
+        return false;
+
+    Ref<API::FrameInfo> frame = API::FrameInfo::create(WTFMove(frameInfo), &page);
```
Super minor nit - we tend to prefer template type deduction when possible.

```suggestion
    Ref frame = API::FrameInfo::create(WTFMove(frameInfo), &page);
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -844,6 +844,50 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
+
+    Ref<API::FrameInfo> frame = API::FrameInfo::create(WTFMove(frameInfo), &page);
+
+    Ref<CompletionHandlerCallChecker> checker = CompletionHandlerCallChecker::create(delegate.get(), @selector(webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:));
```
Ditto

```suggestion
    Ref checker = CompletionHandlerCallChecker::create(delegate.get(), @selector(webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:));
```

lilyspiniolas committed:
[iOS] Allow WKWebView clients to customize the file upload flow for file inputs
https://bugs.webkit.org/show_bug.cgi?id=280537
rdar://130219174

Reviewed by Wenson Hsieh.

These changes allow clients to implement 'webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:'
on iOS if they desire non-default behavior for the file upload flow in
WKWebView. These changes also make WKFileUploadPanel's
'_copyToNewTemporaryDirectory' a class method so that the fix for
rdar://78448610 can be reused here without code duplication.

* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParameters.h:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParameters.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParametersInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKOpenPanelParametersPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::runOpenPanel):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(TestWebKitAPI::TEST(WebKit, RunOpenPanelNonLatin1)):
(TestWebKitAPI::TEST(WebKit, FileInputTypeCancelEvent)):

Canonical link: https://commits.webkit.org/284471@main
